core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -14617,189 +14756,1750 @@ public boolean isUserInRoleWithID(String userID, String roleName) throws UserSto
     public List<Group> listGroups(Condition condition, int limit, int offset, String sortBy,
                                   String sortOrder) throws UserStoreException {
 
-        if (log.isDebugEnabled()) {
-            log.debug("listGroups operation is not implemented in: " + this.getClass());
+        validateCondition(condition);
+        String domain = getDomainFromCondition(condition);
+
+        handlePreListGroup(condition, domain);
+        List<Group> filteredGroups = new ArrayList<>();
+
+        UserStoreManager secManager = getSecondaryUserStoreManager(domain);
+        if (secManager != null) {
+            if (secManager instanceof AbstractUserStoreManager) {
+                if (isUniqueUserIdEnabled(secManager)) {
+                    UniqueIDPaginatedSearchResult <Group> groups = ((AbstractUserStoreManager) secManager)
+                            .doListGroups(condition, domain, offset, limit, sortBy, sortOrder);
+                    filteredGroups = groups.getEntities();
+                } else {
+                    throw new UserStoreException(" listGroups is not supported.");
+                }
+            }
         }
-        throw new NotImplementedException(
-                "listGroups operation is not implemented in: " + this.getClass());
+        handlePostListGroups(condition, domain, false);
+        return filteredGroups;
     }
 
     @Override
     public List<Group> listGroups(Condition condition, int limit, int offset) throws UserStoreException {
 
-        if (log.isDebugEnabled()) {
-            log.debug("listGroups operation is not implemented in: " + this.getClass());
+        validateCondition(condition);
+        String domain = getDomainFromCondition(condition);
+
+        handlePreListGroup(condition, domain);
+        List<Group> filteredGroups = new ArrayList<>();
+
+        UserStoreManager secManager = getSecondaryUserStoreManager(domain);
+        if (secManager != null) {
+            if (secManager instanceof AbstractUserStoreManager) {
+                if (isUniqueUserIdEnabled(secManager)) {
+                    UniqueIDPaginatedSearchResult <Group> groups = ((AbstractUserStoreManager) secManager)
+                            .doListGroups(condition, domain, offset, limit, null, null);
+                    filteredGroups = groups.getEntities();
+                } else {
+                    throw new UserStoreException(" listGroups is not supported.");
+                }
+            }
```

Comment:
Since this block is repeated in above, shall we move it to a separate private method?